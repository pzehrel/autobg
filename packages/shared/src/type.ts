/**
 * A number or a string that can be converted to a number.
 */
export type NumberString = number | string

/**
 * A percentage string.
 */
export type Percentage = `${number}%` | (string & {})

export interface Size {
  width: number
  height: number
}

/**
 * A size with unit.
 *
 * @example
 * ```ts
 * const size: UnitSize = {
 *   width: '100px',
 *   height: '100px',
 * }
 * ```
 */
export interface UnitSize {
  width: `${number}${string}` | (string & {})
  height: `${number}${string}` | (string & {})
}
