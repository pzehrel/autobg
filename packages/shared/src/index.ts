export * from './config'
export * from './css'
export * from './image-size'
export * from './path'
export * from './type'
export * from './utils'
