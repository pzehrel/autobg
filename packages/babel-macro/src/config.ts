import type { Config } from '@autobg/shared'
import { resolveConfig } from '@autobg/shared'

export interface AutobgMacroConfig extends Config {

  /**
   * generate css value unit
   * @default 'px'
   */
  unit?: 'px' | 'rem' | 'vw'

  /**
   * design width, only used when unit is 'vw'
   * @default 750
   */
  designWidth?: number

  /**
   * root element font size, only used when unit is 'rem'
   * @default 100
   */
  rootValue?: number

  /**
   * The number of decimal places to retain during unit conversion
   * @default 5
   */
  unitPrecision?: number
}

export type RequiredAutobgMacroConfig = Required<AutobgMacroConfig>

export const defaultAutobgMacroConfig = resolveConfig<RequiredAutobgMacroConfig>({
  unit: 'px',
  designWidth: 750,
  rootValue: 100,
  unitPrecision: 5,
})

export function defineAutobgMacro(config?: AutobgMacroConfig) {
  return config
}
