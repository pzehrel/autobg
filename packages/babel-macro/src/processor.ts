import type { NodePath, PluginPass } from '@babel/core'
import type * as Babel from '@babel/core'
import type { AutobgMacroConfig } from './config'
import process from 'node:process'
import { createCSS, isHttp, resolveFilepath } from '@autobg/shared'
import { autoCssSize, overwrite } from './utils'

interface ProcessorParams {
  ref: NodePath
  config: Required<AutobgMacroConfig>
  state: PluginPass
  babel: typeof Babel
  aspect?: boolean
  argumentNodes: Babel.NodePath[]
}

export function processor(params: ProcessorParams) {
  const { ref, config, state, babel, aspect, argumentNodes } = params

  const root = state.file.opts.root ?? process.cwd()
  const id = state.file.opts.filename as string

  const csspath = argumentNodes[0].evaluate().value as string
  const filepath = resolveFilepath(csspath, id, root, config)

  const css = createCSS(
    isHttp(csspath) ? csspath : `\${${resolveInjectCode(csspath)}}`,
    filepath,
    {
      aspect,
      side: argumentNodes[1]?.evaluate().value,
      value: argumentNodes[2]?.evaluate().value,
      transformSize: size => autoCssSize(size, config),
    },
  )

  overwrite({ babel, ref: ref.parentPath!, css })
}

function resolveInjectCode(path: string) {
  const vite = `new URL('${path}', import.meta.url).href`
  const webpack = `require('${path}')?.default ?? require('${path}')`
  return `typeof require === 'undefined' ? ${vite} : ${webpack}`
}
