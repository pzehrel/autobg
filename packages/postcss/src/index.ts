import type { Config } from '@autobg/shared'
import type { PluginCreator } from 'postcss'
import process from 'node:process'
import { resolveConfig } from '@autobg/shared'
import { createAutobgAtRule } from './rule'

export interface PostcssAutobgPluginConfig extends Config {
  root?: string
}

export const postcssAutobg: PluginCreator<PostcssAutobgPluginConfig> = (config) => {
  const root = (config?.root || process.cwd()).replace(/\\/g, '/')
  return {
    postcssPlugin: 'autobg',
    // make sure to run before other plugins
    prepare: () => ({
      Once: (node) => {
        // @autobg(<path>)
        // @autobg: url(<path>)
        node.walkAtRules(/^autobg:?/, (rule) => {
          createAutobgAtRule(resolveConfig(config), root)(rule)
        })
      },
    }),
  }
}

postcssAutobg.postcss = true

export default postcssAutobg
