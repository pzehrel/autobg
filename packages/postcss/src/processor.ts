import type { CreateCssOptions, RequiredConfig } from '@autobg/shared'
import type { AtRule } from 'postcss'
import { createCSS, normalizePath, resolveCsspath, resolveFilepath } from '@autobg/shared'

const pathRE = /url\(["']?(.+?)["']?\)/
const scaleRE = /([whs]|width|height|scale)\((\d+(?:\.\d+)?%?)\)/
const aspectRE = /([whs]|width|height|scale)\((\d+(?:\.\d+)?%?)?\)/
const scaleWrongRE = /([whs]|width|height|scale)\(\d+(?:\.\d+)?([^\d%]+)\)/

export function createProcessor(config: RequiredConfig, root: string, aspect?: boolean) {
  return (rule: AtRule) => {
    const parent = rule.parent
    if (!parent || !('selector' in parent) || typeof parent.selector !== 'string') {
      throw rule.error('must be used inside the selector')
    }

    const params = rule.params.trim()
    if (!params) {
      throw rule.error('Parameters cannot be empty')
    }

    const pathMatch = params.match(pathRE)
    if (!pathMatch) {
      throw rule.error('Missing url(<path>) parameter')
    }
    const csspath = normalizePath(pathMatch[1])
    if (!csspath) {
      throw rule.error('Resource file path error')
    }

    const scaleWrongMatch = params.match(scaleWrongRE)
    if (scaleWrongMatch) {
      const [, scale, value, unit] = scaleWrongMatch
      throw rule.error(`"${scale}(${value}${unit})" does not support unit "${unit}", the correct format is "${scale}(${value})" or "${scale}(${value}%)"`)
    }

    const cssOptions: CreateCssOptions = { aspect }
    const scaleMatch = params.match(aspect ? aspectRE : scaleRE)
    if (scaleMatch) {
      const [, scale, value] = scaleMatch
      cssOptions.side = scale
      cssOptions.value = value
    }

    const id = rule.source?.input.file || ''
    const filepath = resolveFilepath(csspath, id, root, config)
    const path = resolveCsspath(filepath, root, config)

    const css = createCSS(path, filepath, cssOptions)
    Object.entries(css).forEach(([key, value]) => {
      parent.prepend({ prop: key, value })
    })

    rule.remove()
  }
}
