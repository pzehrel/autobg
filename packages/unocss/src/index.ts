import type { AutobgUnocssConfig } from './config'
import { resolveConfig } from '@autobg/shared'
import { definePreset } from '@unocss/core'
import { name } from '../package.json'
import { createRules } from './rules'
import { createStore } from './store'
import { transformer } from './transformer'

export type { AutobgUnocssConfig } from './config'

export const presetAutobg = definePreset<AutobgUnocssConfig, object>((config) => {
  const opts = resolveConfig(config)

  const store = createStore()
  store.updateRoot({ configRoot: config?.root })

  return {
    name,
    layer: 'base',
    transformers: [transformer(opts, store)],
    rules: createRules(opts, store),
  }
})
