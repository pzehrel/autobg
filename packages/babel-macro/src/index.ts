import type { AutobgMacroConfig } from './config'
import process from 'node:process'
import { createStyle, isHttp, resolveConfig, resolveFilepath } from '@autobg/shared'
import { createMacro } from 'babel-plugin-macros'
import { defaultAutobgMacroConfig } from './config'
import { autoCssSize, overwrite } from './utils'

export { defineAutobgMacro } from './config'
export type { AutobgMacroConfig } from './config'

export const autobg = createMacro(({ references, state, babel, config }) => {
  references.default.forEach((reference) => {
    if (
      !reference.parentPath
      || reference.parentPath.node.type !== 'CallExpression'
      || reference !== reference.parentPath.get('callee')
    ) {
      return
    }

    let nodes = reference.parentPath.get('arguments')
    if (!Array.isArray(nodes)) {
      nodes = [nodes]
    }

    const opts = resolveConfig(defaultAutobgMacroConfig, config as AutobgMacroConfig)
    const csspath = nodes[0].evaluate().value as string

    const root = state.file.opts.root ?? process.cwd()
    const id = state.file.opts.filename as string

    const filepath = resolveFilepath(csspath, id, root, opts)
    const css = createStyle(
      isHttp(csspath) ? csspath : `\${${resolveInjectCode(csspath)}}`,
      filepath,
      {
        transformSize: size => ({
          width: autoCssSize(size.width, opts),
          height: autoCssSize(size.height, opts),
        }),
      },
    )

    overwrite({ babel, reference, css })
  })
}, { configName: 'autobg' })

export default autobg

function resolveInjectCode(path: string) {
  const vite = `new URL('${path}', import.meta.url).href`
  const webpack = `require('${path}')?.default ?? require('${path}')`
  return `typeof require === 'undefined' ? ${vite} : ${webpack}`
}
