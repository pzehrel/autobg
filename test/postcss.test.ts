import type { Input } from 'postcss'
import postcss from 'postcss'
import { describe, expect, it } from 'vitest'
import { createProcessor } from '../packages/postcss/src/processor'
import { configs, paths } from './util'

function createPostcss(params: string, id: string, selector?: string, aspect?: boolean) {
  const root = postcss.root()
  const atRule = postcss.atRule({ name: aspect ? 'autobg-aspect' : 'autobg', params })

  // mock source file
  atRule.source = { input: { file: id } as Input }

  if (selector) {
    const rule = postcss.rule({ selector })
    rule.append(atRule)
    root.append(rule)
  }
  else {
    root.append(atRule)
  }

  return { root, atRule }
}

describe('postcss', () => {
  for (const [platform, { config, root: cwd, id }] of Object.entries(configs)) {
    for (const { type, path, realpath, exist } of paths) {
      it(`[${platform}] "${path}" should be converted to "${realpath}"`, async () => {
        const processor = createProcessor(config, cwd, false)
        const { root, atRule } = createPostcss(`url(${path})`, id, '.test', false)
        processor(atRule)
        const css = root.toString()

        expect(css).toContain(`background-image: url('${realpath}')`)

        if (exist && type !== 'http') {
          expect(css).toContain(`width: 2px;`)
          expect(css).toContain(`height: 2px;`)
        }

        if (!exist || type === 'http') {
          expect(css).not.toContain(`width: 2px;`)
          expect(css).not.toContain(`height: 2px;`)
        }
      })
    }
  }
})

describe('scaling', async () => {
  const classnames = [
    { name: 'url(/foo.png)', contain: [`aspect-ratio: 2/2`], not: ['width: 2px'] },
    { name: 'url(/foo.png) w(200)', contain: [`aspect-ratio: 2/2`, `width: 200px`] },
    { name: 'url(/foo.png) h(200)', contain: [`aspect-ratio: 2/2`, `height: 200px`] },
    { name: 'url(/foo.png) s(0.78)', contain: [`aspect-ratio: 2/2`, `height: 0.78px`] },
    { name: 'url(/foo.png) scale(78%)', contain: [`aspect-ratio: 2/2`, `height: 78%`] },

    { name: 'url(/foo.png) h()', contain: [`aspect-ratio: 2/2;`, `height: 100%;`] },
    { name: 'url(/foo.png) w(200)', contain: [`aspect-ratio: 2/2;`, `width: 200px;`] },
    { name: 'url(/foo.png) height(200)', contain: [`aspect-ratio: 2/2;`, `height: 200px;`] },
  ]

  for (const { name, contain, not } of classnames) {
    it(`"${name}" \t should be ${contain}`, async () => {
      const { config, root: cwd, id } = configs.posix
      const processor = createProcessor(config, cwd, true)
      const { root, atRule } = createPostcss(name, id, '.test', true)
      processor(atRule)
      const css = root.toString()
      for (const item of contain) {
        expect(css).toContain(item)
      }
      if (not) {
        for (const item of not) {
          expect(css).not.toContain(item)
        }
      }
    })
  }
})
