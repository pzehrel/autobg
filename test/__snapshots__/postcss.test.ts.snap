// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`postcss > should handle multiple @autobg rules 1`] = `
"
      .test1 {
        background-position: center;
        background-size: cover;
        background-repeat: no-repeat;
        background-image: url('./test1.jpg');
      }
      .test2 {
        background-position: center;
        background-size: cover;
        background-repeat: no-repeat;
        background-image: url('./test2.jpg');
      }
    "
`;

exports[`postcss > should transform @autobg rule correctly 1`] = `
"
      .test {
        background-position: center;
        background-size: cover;
        background-repeat: no-repeat;
        background-image: url('./test.jpg');
      }
    "
`;

exports[`postcss autobg plugin > should handle multiple @autobg rules 1`] = `
"
      .test1 {
        background-image: url('./test1.jpg');
        background-size: cover;
        background-position: center;
        background-repeat: no-repeat;
      }
      .test2 {
        background-image: url('./test2.jpg');
        background-size: cover;
        background-position: center;
        background-repeat: no-repeat;
      }
    "
`;

exports[`postcss autobg plugin > should transform @autobg rule correctly 1`] = `
"
      .test {
        background-image: url('./test.jpg');
        background-size: cover;
        background-position: center;
        background-repeat: no-repeat;
      }
    "
`;
